misc/ls-extensions/modules/performance-analyzer-services/src/main/java/io/ballerina/PerformanceAnalyzerNodeVisitor.java

```diff
@@ -0,0 +1,536 @@
+import io.ballerina.compiler.api.symbols.TypeSymbol;
+import io.ballerina.compiler.api.symbols.VariableSymbol;
+import io.ballerina.compiler.syntax.tree.AssignmentStatementNode;
+import io.ballerina.compiler.syntax.tree.BasicLiteralNode;
+import io.ballerina.compiler.syntax.tree.BinaryExpressionNode;
+import io.ballerina.compiler.syntax.tree.BracedExpressionNode;
+import io.ballerina.compiler.syntax.tree.CheckExpressionNode;
+import io.ballerina.compiler.syntax.tree.ExplicitNewExpressionNode;
+import io.ballerina.compiler.syntax.tree.ExpressionNode;
+import io.ballerina.compiler.syntax.tree.ForEachStatementNode;
+import io.ballerina.compiler.syntax.tree.FunctionArgumentNode;
+import io.ballerina.compiler.syntax.tree.FunctionDefinitionNode;
+import io.ballerina.compiler.syntax.tree.IfElseStatementNode;
+import io.ballerina.compiler.syntax.tree.ImplicitNewExpressionNode;
+import io.ballerina.compiler.syntax.tree.ModulePartNode;
+import io.ballerina.compiler.syntax.tree.ModuleVariableDeclarationNode;
+import io.ballerina.compiler.syntax.tree.NamedArgumentNode;
+import io.ballerina.compiler.syntax.tree.NodeVisitor;
+import io.ballerina.compiler.syntax.tree.PositionalArgumentNode;
+import io.ballerina.compiler.syntax.tree.RemoteMethodCallActionNode;
+import io.ballerina.compiler.syntax.tree.SeparatedNodeList;
+import io.ballerina.compiler.syntax.tree.SyntaxKind;
+import io.ballerina.compiler.syntax.tree.VariableDeclarationNode;
+import io.ballerina.compiler.syntax.tree.WhileStatementNode;
+import io.ballerina.component.ActionInvocationNode;
+import io.ballerina.component.EndPointNode;
+import io.ballerina.component.ForStatementNode;
+import io.ballerina.component.IfStatementNode;
+import io.ballerina.component.Node;
+import io.ballerina.projects.Document;
+import io.ballerina.tools.diagnostics.Location;
+import io.ballerina.tools.text.LinePosition;
+import io.ballerina.tools.text.LineRange;
+import io.ballerina.utils.ParserUtil;
+import org.eclipse.lsp4j.Range;
+
+import java.util.HashMap;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Visitor to discover the program structure.
+ *
+ * @since 2.0.0
+ */
+public class PerformanceAnalyzerNodeVisitor extends NodeVisitor {
+
+    public static final long DEFAULT_LOOP_SIZE = 2;
+    public static final String ENDPOINTS_KEY = "endpoints";
+    public static final String ACTION_INVOCATION_KEY = "actionInvocations";
+
+    private final HashMap<LineRange, Object> variableMap;
+    private final HashMap<LineRange, String> referenceMap;
+    private final HashMap<String, EndPointNode> endPointDeclarationMap;
+
+    private final Node startNode;
+    private final SemanticModel model;
+    private final String file;
+    private final Range range;
+    private Node currentNode;
+    private Document document;
+    private boolean withinRange = false;
+
+    public PerformanceAnalyzerNodeVisitor(SemanticModel model, String file, Range range) {
```

Comment:
Better option is to create a context to keep these values and access via context APIs. Check whether we can do the same here. If not that's fine.